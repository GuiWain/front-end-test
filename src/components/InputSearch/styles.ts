import styled, { css, DefaultTheme } from 'styled-components'

import { InputSearchProps } from './'

import { MagnifyingGlass } from '@styled-icons/entypo/'

type IconPositionProps = Pick<InputSearchProps, 'iconPosition'>
type WrapperProps = Pick<InputSearchProps, 'disabled' | 'widthInput'> & {
  error?: boolean
}

export type OptionProps = {
  showHidde?: boolean
}

export const MagnifyingGlassIcon = styled(MagnifyingGlass)``

const wrapperModifiers = {
  error: (theme: DefaultTheme) => css`
    ${InputWrapper} {
      border-color: ${theme.colors.errorMedium};

      &:focus-within {
        border: 0.1rem solid ${theme.colors.errorMedium};
      }
    }

    ${Label} {
      color: ${theme.colors.errorMedium};
    }
  `,
  disabled: (theme: DefaultTheme) => css`
    ${DivOptions},
    ${Options},
    ${Span},
    ${Label},
    ${Input},
    ${Icon} {
      cursor: not-allowed;
      color: ${theme.colors.secondaryMedium};

      &::placeholder {
        color: currentColor;
      }
    }
  `,
  widthInput: (width: number) => css`
    width: ${width + 'rem'};
  `
}

export const Wrapper = styled.div.attrs({
  'data-testid': 'Wrapper'
})<WrapperProps>`
  ${({ theme, error, disabled, widthInput }) => css`
    ${error && wrapperModifiers.error(theme)}
    ${disabled && wrapperModifiers.disabled(theme)}
    ${widthInput && wrapperModifiers.widthInput(widthInput)}
  `}
`

export const InputWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;

    background: ${theme.colors.white} 0% 0% no-repeat padding-box;
    border: 0.1rem solid ${theme.colors.secondaryLight};
    border-radius: ${theme.border.smallRadius};
    padding: 0;

    height: 3.6rem;

    &:focus-within {
      box-shadow: 0rem 0.4rem 0.6rem #00000014;
      border: 0.1rem solid ${theme.colors.infoDark};
    }
  `}
`

export const Input = styled.input<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    color: ${theme.colors.secondaryDark};
    font: normal normal ${theme.font.regular} ${theme.font.sizes.xxsmall} /
      2.8rem ${theme.font.family};

    padding: 0.8rem 0.8rem;

    background: transparent;
    border: 0;
    outline: none;
    width: ${iconPosition === 'right' ? `calc(100% - 2.2rem)` : `100%`};

    &:-webkit-autofill {
      -webkit-box-shadow: 0 0 0 2.4 #eaeaea inset;
      filter: none;
    }
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    font: normal normal ${theme.font.semiBold} ${theme.font.sizes.xsmall} / 4rem
      ${theme.font.family};

    letter-spacing: -0.006rem;
    text-align: left;
    color: ${theme.colors.secondaryDark};
    cursor: pointer;
  `}
`

const iconModifiers = {
  iconLeft: (theme: DefaultTheme) => css`
    order: 0;
    background: ${theme.colors.neutralMedium} 0% 0% no-repeat padding-box;
    color: ${theme.colors.secondaryDarkest};
  `,
  iconRight: (theme: DefaultTheme) => css`
    order: 1;

    background: ${theme.colors.infoDark} 0% 0% no-repeat padding-box;
    color: ${theme.colors.white};
  `
}

export const Icon = styled.div.attrs({
  'data-testid': 'DivIcon'
})<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    display: flex;
    justify-content: center;
    align-items: center;

    height: 100%;
    width: 4.4rem;

    cursor: pointer;

    & > svg {
      width: 1.8rem;
      height: 100%;
    }

    ${iconPosition === 'left'
      ? iconModifiers.iconLeft(theme)
      : iconModifiers.iconRight(theme)}
  `}
`

export const Error = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.errorMedium};
    font-size: ${theme.font.sizes.xxxsmall};
    margin: 0.5rem 0;
  `}
`

export const DivOptions = styled.div`
  ${({ theme }) => css`
    background-color: ${theme.colors.white};
    padding: 0 1.1rem;

    max-height: 20rem;
    overflow-y: scroll;

    box-shadow: 0rem 0.4rem 0.8rem #00000014;
    border: 0.1rem solid ${theme.colors.secondaryLight};
  `}
`

const optionsModifier = {
  show: () => css`
    display: block;
  `,
  hidde: () => css`
    display: none;
  `
}

export const Options = styled.div<OptionProps>`
  ${({ showHidde }) => css`
    background-color: transparent;
    cursor: pointer;

    ${!showHidde && optionsModifier.show()}
    ${showHidde && optionsModifier.hidde()}
  `}
`

export const Span = styled.span`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.xxsmall};
    color: ${theme.colors.secondaryDark};
    letter-spacing: 0rem;
    margin: 0.7rem 0;

    display: flex;
    justify-content: flex-start;
    align-items: center;
    gap: 2rem;

    svg {
      width: 2rem;
      height: 100%;
      color: ${theme.colors.infoDark};
    }
  `}
`
