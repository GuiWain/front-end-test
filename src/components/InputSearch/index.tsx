import * as S from './styles'

import { useState, useEffect, InputHTMLAttributes } from 'react'

export type DisplayOptions = {
  [field: string]: boolean
}

export type InputSearchProps = {
  getValue?: (value: string) => void
  onClickSearch?: (value: string) => void
  label?: string
  name?: string
  initialValue?: string
  widthInput?: number
  iconPosition?: 'left' | 'right'
  placeholder?: string
  disabled?: boolean
  error?: string
  searchAutoComplete?: boolean
  optionsAutoComplete?: string[]
  iconAutoComplete?: React.ReactNode
} & InputHTMLAttributes<HTMLInputElement>

const InputSearch = ({
  iconPosition = 'right',
  label,
  name,
  initialValue = '',
  widthInput,
  placeholder,
  error,
  disabled = false,
  getValue,
  onClickSearch,
  searchAutoComplete = false,
  optionsAutoComplete,
  iconAutoComplete,
  ...props
}: InputSearchProps) => {
  const [value, setValue] = useState(initialValue)
  const [iconPositionModifier, setIconPositionModifier] = useState(iconPosition)
  const [displayOption, setDisplayOption] = useState<DisplayOptions>({})
  const [openOptions, setOpenOptions] = useState(false)

  const onClickIcon = () => {
    !!onClickSearch && onClickSearch(value)
  }

  const onClickSpan = (item: string) => {
    setValue(item)
    setOpenOptions(false)
  }

  const onChange = (value: string) => {
    if (value === '' || value === undefined) setOpenOptions(false)
    else setOpenOptions(true)

    if (searchAutoComplete) {
      const values = value.toLowerCase()

      if (optionsAutoComplete) {
        optionsAutoComplete.forEach((option) => {
          const label = option.toLowerCase()

          if (label.indexOf(values) != -1) {
            setDisplayOption((s) => ({ ...s, [option]: false }))
          } else {
            setDisplayOption((s) => ({ ...s, [option]: true }))
          }
        })
      }
    }
    setValue(value)
  }

  useEffect(() => {
    if (searchAutoComplete) {
      setIconPositionModifier('left')

      if (optionsAutoComplete) {
        optionsAutoComplete.forEach((option) => {
          setDisplayOption((s) => ({ ...s, [option]: false }))
        })
      }
    } else setIconPositionModifier('right')
  }, [searchAutoComplete, optionsAutoComplete])

  useEffect(() => {
    if (getValue) getValue(value)
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [value])

  return (
    <S.Wrapper disabled={disabled} error={!!error} widthInput={widthInput}>
      {!!label && <S.Label htmlFor={name}>{label}</S.Label>}
      <S.InputWrapper>
        <S.Icon iconPosition={iconPositionModifier} onClick={onClickIcon}>
          <S.MagnifyingGlassIcon />
        </S.Icon>

        <S.Input
          type="text"
          onChange={(e) => onChange(e.currentTarget.value)}
          placeholder={placeholder}
          value={value}
          iconPosition={iconPositionModifier}
          disabled={disabled}
          name={name}
          {...(label ? { id: name } : {})}
          {...props}
        />
      </S.InputWrapper>
      {!!optionsAutoComplete && openOptions && (
        <S.DivOptions>
          {optionsAutoComplete.map((item) => (
            <S.Options key={item} showHidde={displayOption[item]}>
              <S.Span onClick={() => onClickSpan(item)}>
                {!!iconAutoComplete && iconAutoComplete} {item}
              </S.Span>
            </S.Options>
          ))}
        </S.DivOptions>
      )}
      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  )
}

export default InputSearch
