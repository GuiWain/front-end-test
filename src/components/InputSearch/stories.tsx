import { Story, Meta } from '@storybook/react'
import InputSearch, { InputSearchProps } from '.'

import { Location } from '@styled-icons/evil/'

import { listStates } from './mockStates'

export default {
  title: 'Inputs/InputSearch',
  component: InputSearch,
  argTypes: {
    onClickSearch: { action: 'changed' },
    placeholder: {
      type: 'string'
    },
    label: {
      type: 'string'
    },
    name: {
      type: 'string'
    },
    initialValue: {
      type: 'string'
    },
    widthInput: {
      type: 'number'
    },
    disabled: {
      control: {
        type: 'radio',
        options: [true, false]
      }
    },
    searchAutoComplete: {
      control: {
        type: 'radio',
        options: [true, false]
      }
    },
    error: {
      type: 'string'
    }
  }
} as Meta

const onSearch = (value: string) => {
  alert('Value input: ' + value)
}

export const Default: Story<InputSearchProps> = (args) => (
  <div style={{ width: '30%', margin: ' 2rem 0' }}>
    <InputSearch {...args} />
  </div>
)

Default.args = {
  label: 'Search by name',
  name: 'name',
  placeholder: 'Search by name',
  disabled: false,
  onClickSearch: onSearch,
  searchAutoComplete: false
}

export const AutoComplete: Story<InputSearchProps> = (args) => (
  <div style={{ width: '30%', margin: ' 2rem 0' }}>
    <InputSearch {...args} />
  </div>
)

AutoComplete.args = {
  label: 'Search by state',
  name: 'name',
  placeholder: 'Search by city or abbreviation',
  disabled: false,
  searchAutoComplete: true,
  optionsAutoComplete: listStates,
  iconAutoComplete: <Location />
}

export const WithError: Story<InputSearchProps> = (args) => (
  <div style={{ width: '30%', margin: ' 2rem 0' }}>
    <InputSearch {...args} />
  </div>
)

WithError.args = {
  label: 'Search',
  name: 'name',
  placeholder: 'Search by name',
  disabled: false,
  error: 'Required field'
}
