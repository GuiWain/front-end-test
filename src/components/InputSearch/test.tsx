import userEvent from '@testing-library/user-event'
import { render, screen } from 'utils/test-utils'

import { listStates } from './mockStates'
import { Location } from '@styled-icons/evil/'

import InputSearch from '.'

describe('<InputSearch />', () => {
  it('should render the InputSearch', () => {
    const { container } = render(<InputSearch />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('Renders with label', () => {
    render(<InputSearch label="labelText" name="nameText" />)

    expect(screen.getByLabelText('labelText')).toBeInTheDocument()
  })

  it('Renders with initialValue', () => {
    const text = 'This is my new text'
    render(<InputSearch initialValue={text} />)

    const input = screen.getByRole('textbox')
    expect(input).toHaveValue(text)
  })

  it('Renders with placeholder', () => {
    render(<InputSearch placeholder="hey you" />)

    expect(screen.getByPlaceholderText('hey you')).toBeInTheDocument()
  })

  it('Renders with error', () => {
    render(<InputSearch error="Required field" />)

    expect(screen.getByText('Required field')).toBeInTheDocument()
  })

  it('Render with disabled', () => {
    render(<InputSearch label="InputSearch" name="InputSearch" disabled />)

    const input = screen.getByRole('textbox')
    expect(input).toBeDisabled()
  })

  it('Renders with getValue', () => {
    const getValue = jest.fn()

    render(
      <InputSearch label="InputSearch" name="InputSearch" getValue={getValue} />
    )

    const input = screen.getByRole('textbox')
    const text = 'This is my new text'
    userEvent.type(input, text)

    expect(getValue).toHaveBeenCalled()
  })

  it('Renders with widthInput', () => {
    render(
      <InputSearch label="InputSearch" name="InputSearch" widthInput={45} />
    )

    expect(screen.getByTestId('Wrapper')).toHaveStyle({
      width: '45rem'
    })
  })

  it('Renders with onClickSearch', () => {
    const onClickSearch = jest.fn()

    render(
      <InputSearch
        label="InputSearch"
        name="InputSearch"
        onClickSearch={onClickSearch}
      />
    )

    const input = screen.getByRole('textbox')
    const text = 'This is my new text'
    userEvent.type(input, text)

    const icon = screen.getByTestId('DivIcon')
    userEvent.click(icon)

    expect(onClickSearch).toHaveBeenCalled()
  })

  it('Renders with autoComplete', () => {
    render(
      <InputSearch
        label="InputSearch"
        name="InputSearch"
        searchAutoComplete={true}
      />
    )

    const input = screen.getByRole('textbox')
    const text = 'This is my new text'
    userEvent.type(input, text)

    expect(input).toHaveValue(text)
  })

  it('Renders with optionsAutoComplete and iconAutoComplete', () => {
    render(
      <InputSearch
        label="InputSearch"
        name="InputSearch"
        searchAutoComplete={true}
        optionsAutoComplete={listStates}
        iconAutoComplete={<Location />}
      />
    )

    expect(screen.queryByText(/Acre/i)).not.toBeInTheDocument()

    const input = screen.getByRole('textbox')
    const text = 'AC'
    userEvent.type(input, text)
    expect(screen.queryByText(/Acre/i)).toBeInTheDocument()

    const span = screen.getByText(/Acre/i)
    userEvent.click(span)
    expect(screen.queryByText(/Alagoas/i)).not.toBeInTheDocument()

    userEvent.clear(input)
    expect(screen.queryByText(/Acre/i)).not.toBeInTheDocument()
  })
})
