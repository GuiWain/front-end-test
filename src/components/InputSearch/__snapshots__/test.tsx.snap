// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputSearch /> should render the InputSearch 1`] = `
.c2 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFFFFF 0% 0% no-repeat padding-box;
  border: 0.1rem solid #CCC8C8;
  border-radius: 0.4rem;
  padding: 0;
  height: 3.6rem;
}

.c0:focus-within {
  box-shadow: 0rem 0.4rem 0.6rem #00000014;
  border: 0.1rem solid #3241A1;
}

.c3 {
  color: #4F4C4C;
  font: normal normal 400 1.4rem / 2.8rem 'Open Sans','Roboto','Helvetica Neue',sans-serif;
  padding: 0.8rem 0.8rem;
  background: transparent;
  border: 0;
  outline: none;
  width: calc(100% - 2.2rem);
}

.c3:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 2.4 #eaeaea inset;
  -webkit-filter: none;
  filter: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  width: 4.4rem;
  cursor: pointer;
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
  background: #3241A1 0% 0% no-repeat padding-box;
  color: #FFFFFF;
}

.c1 > svg {
  width: 1.8rem;
  height: 100%;
}

<div
  class=""
  data-testid="Wrapper"
>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="DivIcon"
    >
      <svg
        aria-hidden="true"
        class="c2 "
        fill="currentColor"
        focusable="false"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17.545 15.467l-3.779-3.779a6.15 6.15 0 00.898-3.21c0-3.417-2.961-6.377-6.378-6.377A6.185 6.185 0 002.1 8.287c0 3.416 2.961 6.377 6.377 6.377a6.15 6.15 0 003.115-.844l3.799 3.801a.953.953 0 001.346 0l.943-.943c.371-.371.236-.84-.135-1.211zM4.004 8.287a4.282 4.282 0 014.282-4.283c2.366 0 4.474 2.107 4.474 4.474a4.284 4.284 0 01-4.283 4.283c-2.366-.001-4.473-2.109-4.473-4.474z"
        />
      </svg>
    </div>
    <input
      class="c3"
      type="text"
      value=""
    />
  </div>
</div>
`;
